### DISCUSSION TITLE
Changed caching behavior of permanent packages on uninstall

### DISCUSSION THREAD

**Discussion by ssrmm** [wixtoolset/issues#8543] [Questions]
I've recently updated a WiX based installer from v3 to v5. We are seeing the following behavioral difference for which I'm not sure if it is intended or a bug:

We install some `ExePackage`s in a burn bundle. They are prerequisites that need to stay installed after uninstall because they are shared components and may still be used by other, unrelated applications on the system. So we use `Permanent="yes"` for them. We currently don't specify any cache-behavior, so they are using the default `Cache="keep"`. This has behaved as expected in WiX v3, caching the packages on install and removing them from the cache on uninstall. In WiX v5 however, they remain in the cache even after uninstall.

It seems there ware some changes to caching behavior as written [here](https://www.firegiant.com/blog/2021/3/18/wix-online-meeting-208-highlights/) and [here](https://www.firegiant.com/blog/2021/4/15/wix-online-meeting-210-highlights/). From the [corresponding documentation](https://wixtoolset.org/docs/schema/wxs/keepremoveforcetype/) it sounds like what I'm seeing might be intentional.

This seems weird to me however, because after the bundle is uninstalled, there is nothing referencing the packages anymore (unless of course you reinstall the same version again) and I don't understand for what reason they would need to be kept. So to me it feels like they are just wasting disk space at that point. I haven't tested yet, but I guess these would also accumulate in multiple versions if you upgrade through multiple versions of the bundle.

Note that using `Cache="remove"` is not helpful for our use case either. We still want to be able to have _Repair_ repair accidentally uninstalled prerequisites, With `Cache="remove"` it can't do that because it is missing the package at the time of repair.

Can somebody clarify if this is intentional behavior or a bug? And if it is intentional: What is the use-case being supported by it?

**Comment by ssrmm** [comment]
While looking into some other issue, I found out an interesting piece of information: This behavior is influenced by `DetectCondition`: The cached prerequisites are removed if and only if `DetectCondition` evaluates to `false`. In the installer in question, `DetectCondition` will usually evaluate to `true` during uninstallation. So effectively there are two cases, depending on the order in which things get uninstalled:

**Case 1:**
1. The bundle is installed and burn installs the prerequisite package as part of that.
2. The prerequisite is uninstalled manually by the user.
3. The bundle is uninstalled manually by the user.

In this case the `DetectCondition` evaluates `false` at the time the bundle is uninstalled and everything is cleaned up and the `Package Cache` folder ends up in the same state as it was before step one.

**Case 2:**
1. The bundle is installed and burn installs the prerequisite package as part of that.
2. The bundle is uninstalled manually by the user (but burn does not uninstall the prerequisite because it is `Permanent`).
3. Optional: The prerequisite is uninstalled manually by the user.

In this case the `DetectCondition` evaluates `true` at the time the bundle is uninstalled and the installer of the prerequisite stays around in the `Package Cache` folder and presumably never gets cleaned up. I feel like this shouldn't be the case. To my understanding the leftover prerequisite packages in the `Package Cache` folder "belong" to the bundle, so they should be removed with it.

@robmen @barnson Any chance either of you can shed some light on whether this is intentional or not?

**Reply by barnson** [reply]
Dunno. All that logic is in plan.cpp; maybe there's a comment suggesting a reason?
